Some changes to the vimrc

chrisbra opened the pull request:
Hi this branch changes some of the Vim settings. I tried to make the commits as granular as possible to make reviewing easier.

Currently some settings can be improved a bit, so I slightly changed them. I tried not to change them according to my personal taste and if I did, I mentioned it in the commit message (cursor block style, command line completion, ...) and those could then reverted if needed.

The rest should keep Vim more closely to Vims upstream repository, especially since it sources the defaults.vim file anyhow (currently overwriting some settings). 

Also I found comments beneficial, so I left a couple of more comments. One thing I did not know for sure is the usage of EDIT_MERGEMSG and EDIT_MSG. I am not sure whether both of them are needed or only one of them (and the other one was a typo). For now I left it as is (and added the other one to the settings as well. This can easily be reverted if needed. Also if this is correct, it might be beneficial to add some of this to upstreams filetype detection logic.

dscho commented on git-extra/vimrc:
```diff
@@ -19,17 +19,16 @@ set statusline=%<%F%h%m%r\ [%{&ff}]\ (%{strftime(\"%H:%M\ %d/%m/%Y\",getftime(ex
       \           && expand("%") !~ "ADD_EDIT.patch"
-      \           && expand("%") !~ "addp-hunk-edit.diff"
-      \           && expand("%") !~ "git-rebase-todo" |
+      \           && expand("%") !~ "addp-hunk-edit.diff" |
```
For somebody versed in vim's command language, this might all look very clear. To everybody else, this is just magic.

That includes me, and I need to understand those changes before I can merge the PR. Please take a little bit of time to explain in the commit message what you did there, at least enough for somebody e.g. stuck in an airplane without internet access (and therefore unable to look up documentation and things) to  be able to change the `&ft` line to their needs.

dscho commented on git-extra/vimrc:
```diff
@@ -20,13 +20,13 @@ set statusline=%<%F%h%m%r\ [%{&ff}]\ (%{strftime(\"%H:%M\ %d/%m/%Y\",getftime(ex
 if has("autocmd")
     " Set UTF-8 as the default encoding for commit messages
-    autocmd BufReadPre COMMIT_EDITMSG,MERGE_MSG,git-rebase-todo setlocal fileencodings=utf-8
+    autocmd BufReadPre COMMIT_EDITMSG,MERGE_EDITMSG,MERGE_MSG,git-rebase-todo setlocal fileencodings=utf-8
```
This comes from https://github.com/git-for-windows/build-extra/commit/f5cc31f85522b92d0ee1e7ea11de2234cd7cc37d, which references https://github.com/git-for-windows/git/issues/1207 as reason for introducing the `MERGE_EDITMSG` file name here.

As you can see easily from reading that discussion, I simply made a mistake, and `MERGE_MSG` was really intended.

Maybe take this opportunity to fix it?

chrisbra replied:
Sure, will do.

dscho commented on git-extra/vimrc:
```diff
@@ -37,6 +37,4 @@ if has("autocmd")
       \ highlight WhiteSpaceEOL ctermbg=red |
       \ match WhiteSpaceEOL /\(^+.*\)\@<=\s\+$/
-
-      autocmd Syntax gitcommit setlocal textwidth=74
```
Yep, this stems from a time when this was *not* configured automatically.

chrisbra replied:
I am pretty sure, Vim has support for the diff filetype for more than 15 years however. But most likely, filetype detection was not enabled back then by default.

dscho replied:
Ah! Thanks for the clarification. I misunderstood.

dscho commented on git-extra/vimrc:
```diff
@@ -14,7 +14,16 @@ set clipboard=unnamed           " set clipboard to unnamed to access the system
+set statusline+=\ (%{strftime(\"%H:%M\ %d/%m/%Y\",getftime(expand(\"%:p\")))})  " last modified timestamp
+set statusline+=%=              " Rest: right align
+set statusline+=%l,%c%V         " Position in buffer: linenumber, column, virtual column
+set statusline+=\ %P            " Position in buffer: Percentage
```
Nicely done!

Since this is *expanding* rather than *simplifying*, let's call it that in the commit message?

Also, I see that `%<` was removed. Can you describe in the commit message, please, what this does and why it is okay to remove it?

chrisbra replied:
Sure, will do.

dscho replied:
Thanks!

dscho commented on git-extra/vimrc:
```diff
@@ -25,6 +25,12 @@ set statusline+=%=              " Rest: right align
 set statusline+=\ %P            " Position in buffer: Percentage
 
+if &term =~ 'xterm'
+  let &t_ti.="\e[1 q"           " Use a blinking block cursor when entering termcap mode
```
Does this let the cursor blink by default?

I wonder how that interacts with MinTTY. I configured a block cursor there, and I could imagine that other users did so, too. Yet other users may have configured a line cursor specifically and could possibly be pretty annoyed by this here patch.

What's your take on this?

dscho commented on git-extra/vimrc:
```diff
@@ -30,6 +30,12 @@ if &term =~ 'xterm'
+  endif
+  if 0
+    set termguicolors           " Allow trucolors on mintty 
+  endif
```
What is the effect of this? The commit message stops short of explaining that.

Also, the commit message starts with an unnecessarily controversial "better". Let's avoid that. Let's just call it "Use more colors".

chrisbra replied:
will do

dscho commented:
I reviewed all patches, and left a few comments and suggestions.

chrisbra commented on git-extra/vimrc:
```diff
@@ -1,16 +1,16 @@
 
+ru! defaults.vim                " Use Enhanced Vim defaults
+set mouse=                      " Reset the mouse setting from defaults
+aug vimStartup | au! | aug END  " Revert last positioned jump, as it is defined below
```
Of course, this is Vim :)

chrisbra commented on git-extra/vimrc:
```diff
@@ -19,17 +19,16 @@ set statusline=%<%F%h%m%r\ [%{&ff}]\ (%{strftime(\"%H:%M\ %d/%m/%Y\",getftime(ex
     autocmd BufReadPre COMMIT_EDITMSG,MERGE_MSG,git-rebase-todo setlocal fileencodings=utf-8
 
-    "Remember the positions in files with some git-specific exceptions"
+    " Remember the positions in files with some git-specific exceptions"
```
correct

dscho replied:
In that case, could I ask to split them out? I was really confused because I am not familiar with the Vim syntax, so I thought you introduced functional changes here that I do not understand.

dscho replied:
Ah, I see that you explained that part in the commit message. No need to split anything anymore!

chrisbra commented:
> Could you please add "Signed-off-by:" footers to the commit messages (that's why the DCO bot marks this PR with a red X)?

Yeah I saw that, once I created the PR. If I had been a bit more careful before creating the PR, I would have Signed-off all those commits. I usually do not follow the git practice in this regard.

But since I had to reword (and slightly change) some of the commits anyhow, I also added Signed-off-by header. 

Also I reworded the commits as required to explain a bit, why something was changed (and dropped the use blinking cursor commit, that was unneccessary and I should have configured my terminal directly instead of trying to fix this in Vim.)

dscho commented on git-extra/vimrc:
```diff
@@ -11,6 +11,7 @@ set vb                          " turn on the "visual bell" - which is much quie
 set laststatus=2                " make the last line where the status is two lines deep so you can see status always
 set showmode                    " show the current mode
 set clipboard=unnamed           " set clipboard to unnamed to access the system clipboard under windows
+set wildmode=list:longest,longest:full   " Better command line completion
```
Thank you, the new commit message does a really nice job of explaining it!

dscho commented:
Thank you for your contribution, and for your patience with me!

chrisbra commented:
Thanks for merging. One minor comment

> I try to make a point in this project to be inclusive. Next time, please include the roughly 50% of humanity that happens to have a different sex than you do.

Hm I always thought he refers to `the user` which in my understanding would be neither masculine nor feminin. In German I don't think of 'männlich' wenn talking about "der Anwender". But it's not my monther tongue as you might guess, so I beg for pardon for the misunderstanding.

dscho commented:
> I always thought he refers to the user which in my understanding would be neither masculine nor feminin

No, in English "he" refers to a male being. Imagine how you would feel if you were addressed as a "she" 100% of the time? That's exactly why this should be avoided. Hence the "singular they". Example:

> Whenever a user wants to open a file using Git's default editor, they only need to call `git config -e -f <path>`.
